File: Sudoku.ml
open Printf

let flip f x y = f y x

let range n =
  let rec loop i xs =
    if i = 0 then
      xs
    else
      loop (i - 1) ((i - 1) :: xs)
  in
  loop n []

let minimum_with_index arr =
  let l = Array.length arr in
  let rec loop (i, m) j =
    if j = l then
      (i, m)
    else if arr.(j) < m then
      loop (j, arr.(j)) (j + 1)
    else
      loop (i, m) (j + 1)
  in
  if l = 0 then
    failwith "minimum_with_index called on empty array"
  else
    loop (0, arr.(0)) 1

let read_line_option () =
  try Some (read_line ())
  with End_of_file -> None

type sudoku = { admit : bool array; count : int array }

exception No_solution

let box_size = 3
let size     = box_size * box_size
let grd      = 0
let row      = 1
let col      = 2
let box      = 3
let view     = 4
let n_coord  = size * size * size
let n_block  = view * size * size
let defined  = 0xDEF
let digits   = "123456789"

let coord i j k = (i * size + j) * size + k
let block v p q = (v * size + p) * size + q

let parents =
  Array.init n_coord (fun c ->
    let i = c / size / size in
    let j = c / size mod size in
    let k = c mod size in
    let p = i / box_size * box_size + j / box_size in
    [block grd i j; block row i k; block col j k; block box p k]
  )

let children =
  let children' = Array.make n_block [] in
  let () =
    flip Array.iteri parents (fun c bs ->
      flip List.iter bs (fun b ->
        children'.(b) <- c :: children'.(b)
      )
    )
  in
  Array.init n_block (fun c -> List.rev children'.(c))

let init () =
  { admit = Array.make n_coord true; count = Array.make n_block size }

let copy s =
  { admit = Array.copy s.admit; count = Array.copy s.count }

let assign s c =
  let q = Queue.create () in
  let () = Queue.add c q in
  let rec loop () =
    if Queue.is_empty q then
      s
    else (
      let c0 = Queue.take q in
      if not s.admit.(c0) then raise No_solution;
      flip List.iter parents.(c0) (fun b1 ->
        s.count.(b1) <- defined;
        flip List.iter children.(b1) (fun c2 ->
          if c2 <> c0 && s.admit.(c2) then (
            s.admit.(c2) <- false;
            flip List.iter parents.(c2) (fun b3 ->
              s.count.(b3) <- s.count.(b3) - 1;
              if s.count.(b3) = 0 then raise No_solution;
              if s.count.(b3) = 1 then (
                flip List.iter children.(b3) (fun c4 ->
                  if s.admit.(c4) then Queue.add c4 q
                )
              )
            )
          )
        )
      );
      loop ()
    )
  in
  loop ()

let read input' =
  let s = init () in
  let input =
    if String.length input' > size * size then
      String.sub input' 0 (size * size)
    else
      input'
  in
  let () =
    flip String.iteri input (fun ij ch ->
      try
        let i = ij / size in
        let j = ij mod size in
        let k = String.index digits ch in
        ignore (assign s (coord i j k))
      with Not_found -> ()
    )
  in
  s

let show s =
  let buf = Buffer.create (size * size) in
  let () =
    flip List.iter (range size) (fun i ->
      flip List.iter (range size) (fun j ->
        match List.filter (fun k -> s.admit.(coord i j k)) (range size) with
        | [k] -> Buffer.add_char buf (String.get digits k)
        | _   -> Buffer.add_char buf '.'
      )
    )
  in
  Buffer.contents buf

let rec search s =
  let (b, m) = minimum_with_index s.count in
  if m = defined then
    s
  else
    let rec loop = function
      | [] ->
          raise No_solution
      | c :: cs ->
          try search (assign (copy s) c)
          with No_solution -> loop cs
    in
    loop children.(b)

let solve input =
  try show (search (read input))
  with No_solution -> "NO SOLUTION"

let () =
  let rec loop () =
    match read_line_option () with
    | None       -> ()
    | Some input -> printf "%s\n%!" (solve input); loop ()
  in
  loop ()

File: Sudoku.rs
use std::collections::VecDeque;

fn main() {
    use std::io::BufRead;
    let rule = Rule::new();
    let stdin = std::io::stdin();
    for problem in stdin.lock().lines().filter_map(|r| r.ok()) {
        println!("{}", Sudoku::solve(problem, &rule));
    }
}

struct Rule {
    size: usize,
    digits: String,
    n_coord: usize,
    n_block: usize,
    parents: Vec<Vec<Block>>,
    children: Vec<Vec<Coord>>,
}

type Block = usize;
type Coord = usize;

impl Rule {
    fn new() -> Self {
        let unit = 3;
        let size = unit * unit;
        let digits = "123456789".to_string();
        assert_eq!(size, digits.len());
        let n_coord = size * size * size;
        let n_block = 4 * size * size;
        let mut parents = Vec::with_capacity(n_coord);
        for i in 0..size {
            for j in 0..size {
                let p = i / unit * unit + j / unit;
                for k in 0..size {
                    parents.push(vec![
                        (0 * size + i) * size + j,
                        (1 * size + i) * size + k,
                        (2 * size + j) * size + k,
                        (3 * size + p) * size + k,
                    ]);
                }
            }
        }
        let mut children = vec![Vec::with_capacity(size); n_block];
        for c in 0..n_coord {
            for &b in &parents[c] {
                children[b].push(c);
            }
        }
        Rule {
            size,
            digits,
            n_coord,
            n_block,
            parents,
            children,
        }
    }
}

#[derive(Clone)]
struct Sudoku<'a> {
    state: Vec<State>,
    count: Vec<Count>,
    queue: VecDeque<Coord>,
    rule: &'a Rule,
}

#[derive(Clone, PartialEq)]
enum State {
    Vague,
    False,
    True,
}

type Count = usize;

const COUNT_DONE: Count = std::usize::MAX;

impl<'a> Sudoku<'a> {
    fn solve(problem: String, rule: &'a Rule) -> String {
        let mut s = Sudoku::new(rule);
        s.feed(problem);
        if let Some(s) = s.search() {
            s.display()
        } else {
            "NO SOLUTION".to_string()
        }
    }
    fn new(rule: &'a Rule) -> Self {
        Sudoku {
            state: vec![State::Vague; rule.n_coord],
            count: vec![rule.size; rule.n_block],
            queue: VecDeque::new(),
            rule,
        }
    }
    fn feed(&mut self, problem: String) {
        let size = self.rule.size;
        for (ij, ch) in problem.chars().take(size * size).enumerate() {
            let i = ij / size;
            let j = ij % size;
            if let Some(k) = self.rule.digits.find(ch) {
                self.queue.push_back((i * size + j) * size + k);
            }
        }
    }
    fn display(&self) -> String {
        let size = self.rule.size;
        let mut res = String::with_capacity(size * size);
        for i in 0..size {
            for j in 0..size {
                res.push(
                    (0..size)
                        .find(|&k| self.state[(i * size + j) * size + k] == State::True)
                        .map_or('.', |k| self.rule.digits.chars().nth(k).unwrap()),
                );
            }
        }
        res
    }
    fn search(mut self) -> Option<Self> {
        while let Some(c) = self.queue.pop_front() {
            if !self.assign(c) {
                return None;
            }
        }
        let mut b_min = 0;
        let mut min = self.count[b_min];
        for b in 1..self.rule.n_block {
            if self.count[b] < min {
                b_min = b;
                min = self.count[b_min];
            }
        }
        if min == COUNT_DONE {
            return Some(self);
        }
        self.rule.children[b_min].iter().find_map(|&c| {
            let mut s = self.clone();
            s.queue.push_back(c);
            s.search()
        })
    }
    fn assign(&mut self, c: Coord) -> bool {
        match self.state[c] {
            State::Vague => {
                self.state[c] = State::True;
                self.rule.parents[c].iter().all(|&b| self.mark(b))
            }
            State::False => false,
            State::True => true,
        }
    }
    fn mark(&mut self, b: Block) -> bool {
        self.count[b] = COUNT_DONE;
        self.rule.children[b].iter().all(|&c| self.ban(c))
    }
    fn ban(&mut self, c: Coord) -> bool {
        match self.state[c] {
            State::Vague => {
                self.state[c] = State::False;
                self.rule.parents[c].iter().all(|&b| self.countdown(b))
            }
            _ => true,
        }
    }
    fn countdown(&mut self, b: Block) -> bool {
        if self.count[b] == COUNT_DONE {
            return true;
        }
        self.count[b] -= 1;
        if self.count[b] == 0 {
            return false;
        }
        if self.count[b] == 1 {
            for &c in &self.rule.children[b] {
                if self.state[c] == State::Vague {
                    self.queue.push_back(c);
                }
            }
        }
        return true;
    }
}
